cpp/src/quantiles/quantiles.cu

```diff
@@ -20,60 +20,104 @@
+                          gdf_size_type n,
+                          double q, 
+                          gdf_quantile_method interpolation,
+                          RetT& result,
+                          bool flag_sorted = false,
+                          cudaStream_t stream = NULL)
   {
-    using FctrType = std::function<RetT(VType, VType, double)>;
+    std::vector<T> hv(2);
 
-    FctrType lin_interp{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(static_cast<double>(y0) + x*static_cast<double>(y1-y0));//(f(x) - y0) / (x - 0) = m = (y1 - y0)/(1 - 0)
-      }};
+    if( q >= 1.0 && flag_sorted )
+    {
+      T* d_res = thrust::max_element(rmm::exec_policy(stream)->on(stream), dv, dv+n);
+      cudaMemcpy(&hv[0], d_res, sizeof(T), cudaMemcpyDeviceToHost);//TODO: async with streams?
+      result = static_cast<RetT>( hv[0] );
+      return GDF_SUCCESS;
+    }
 
-    FctrType midpoint{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(static_cast<double>(y0 + y1)/2.0);
-      }};
+    if( q <= 0.0 && flag_sorted )
+    {
+      T* d_res = thrust::min_element(rmm::exec_policy(stream)->on(stream), dv, dv+n);
+      cudaMemcpy(&hv[0], d_res, sizeof(T), cudaMemcpyDeviceToHost);//TODO: async with streams?
+      result = static_cast<RetT>( hv[0] );
+      return GDF_SUCCESS;
+    }
 
-    FctrType nearest{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(x < 0.5 ? y0 : y1);
-      }};
+    if( n < 2 )
+    {
+      cudaMemcpy(&hv[0], dv, sizeof(T), cudaMemcpyDeviceToHost);//TODO: async with streams?
+      result = static_cast<RetT>( hv[0] );
+      return GDF_SUCCESS;
+    }
 
-    FctrType lowest{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(y0);
-      }};
+    // sort if the input is not sorted.
+    if( !flag_sorted ){
+      thrust::sort(rmm::exec_policy(stream)->on(stream), dv, dv+n);
+    }
 
-    FctrType highest{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(y1);
-      }};
-  
-  
-    std::vector<std::string> methods{"lin_interp", "midpoint", "nearest", "lowest", "highest"};
-    size_t n_methods = methods.size();
-    std::vector<FctrType> vf{lin_interp, midpoint, nearest, lowest, highest};
-  
-    std::vector<double> qvals{0.0, 0.25, 0.33, 0.5, 1.0};
+    QuantiledIndex qi = find_quantile_index(n, q);
 
-  
-    assert( n_methods == methods.size() );
-  
-    for(auto q: qvals)
-      {
-        VType res = quantile_approx(d_in.data().get(), d_in.size(), q);
-        std::cout<<"q: "<<q<<"; exact res: "<<res<<"\n";
-        for(auto i = 0;i<n_methods;++i)
-          {
-            RetT rt = quantile_exact(d_in.data().get(), d_in.size(), q, vf[i]);
-            std::cout<<"q: "<<q<<"; method: "<<methods[i]<<"; rt: "<<rt<<"\n";
-          }
-      }
-  }
+    switch( interpolation )
+    {
+    case GDF_QUANT_LINEAR:
+      cudaMemcpy(&hv[0], dv+qi.lower_bound, sizeof(T), cudaMemcpyDeviceToHost); //TODO: async with streams

```

Comment:
All of these `cudaMemcpy`s need to be wrapped in `CUDA_TRY`.

Reply:
In anticipation of using streams, all of these should be made async and accept the defaulted `stream` parameter passed to this function.